File: packages/server/src/core/db/schema.ts
import { sqliteTable, text, integer, primaryKey } from 'drizzle-orm/sqlite-core'

// ── Projects ──────────────────────────────────────────────────────────────────

export const projects = sqliteTable('projects', {
  id: text('id').primaryKey(),
  name: text('name').notNull(),
  color: text('color').notNull().default('#6366f1'),
  createdAt: integer('created_at').notNull()
})

// ── Project ↔ Task relationship ───────────────────────────────────────────────

export const projectTasks = sqliteTable(
  'project_tasks',
  {
    projectId: text('project_id').notNull().references(() => projects.id, { onDelete: 'cascade' }),
    taskId: text('task_id').notNull().references(() => tasks.id, { onDelete: 'cascade' }),
    addedAt: integer('added_at').notNull()
  },
  (t) => ({
    pk: primaryKey({ columns: [t.projectId, t.taskId] })
  })
)

// ── Tasks ─────────────────────────────────────────────────────────────────────

export const tasks = sqliteTable('tasks', {
  id: text('id').primaryKey(),
  parentTaskId: text('parent_task_id'),
  agentName: text('agent_name'),
  agentRole: text('agent_role'),
  prompt: text('prompt').notNull(),
  attachments: text('attachments'), // JSON string[]
  status: text('status').notNull().default('pending_approval'),
  mode: text('mode').notNull().default('browser'),
  capabilities: text('capabilities').notNull(), // JSON TaskCapabilityId[]
  steps: text('steps').notNull().default('[]'), // JSON TaskStep[]
  usage: text('usage'), // JSON { inputTokens, outputTokens }
  createdAt: integer('created_at').notNull(),
  updatedAt: integer('updated_at').notNull(),
  maxSteps: integer('max_steps').notNull(),
  timeoutMs: integer('timeout_ms').notNull(),
  error: text('error'),
  summary: text('summary'),
  source: text('source')
})

// ── Schedules ─────────────────────────────────────────────────────────────────

export const schedules = sqliteTable('schedules', {
  id: text('id').primaryKey(),
  prompt: text('prompt').notNull(),
  capabilities: text('capabilities').notNull(), // JSON TaskCapabilityId[]
  mode: text('mode').notNull(),
  frequency: text('frequency').notNull(),
  cronExpr: text('cron_expr').notNull(),
  enabled: integer('enabled', { mode: 'boolean' }).notNull().default(true),
  lastRunAt: integer('last_run_at'),
  nextRunAt: integer('next_run_at').notNull(),
  createdAt: integer('created_at').notNull()
})

// ── Skills ────────────────────────────────────────────────────────────────────

export const skills = sqliteTable('skills', {
  id: text('id').primaryKey(),
  name: text('name').notNull(),
  tag: text('tag').notNull(),
  description: text('description').notNull(),
  prompt: text('prompt').notNull(),
  enabled: integer('enabled', { mode: 'boolean' }).notNull().default(true),
  createdAt: integer('created_at').notNull()
})

// ── Secrets ───────────────────────────────────────────────────────────────────

export const secrets = sqliteTable('secrets', {
  key: text('key').primaryKey(),
  value: text('value').notNull(),
  updatedAt: integer('updated_at').notNull()
})

// ── Policy (single-row config) ────────────────────────────────────────────────

export const policy = sqliteTable('policy', {
  id: integer('id').primaryKey({ autoIncrement: true }),
  workspaceRoot: text('workspace_root'),
  capabilities: text('capabilities').notNull(), // JSON Record<CapabilityId, boolean>
  themeMode: text('theme_mode').notNull().default('dark'),
  language: text('language').notNull().default('en'),
  providerActive: text('provider_active').notNull().default('chatgpt'),
  providerOpenaiModel: text('provider_openai_model').notNull().default('gpt-4.1-mini'),
  taskMemoryEnabled: integer('task_memory_enabled', { mode: 'boolean' }).notNull().default(true),
  taskAutoApprove: integer('task_auto_approve', { mode: 'boolean' }).notNull().default(false)
})

File: packages/core/src/schema.rs
diesel::table! {
    tasks (id) {
        id -> Text,
        prompt -> Text,
        attachments -> Nullable<Text>,
        status -> Text,
        mode -> Text,
        capabilities -> Text,
        steps -> Text,
        usage -> Nullable<Text>,
        created_at -> BigInt,
        updated_at -> BigInt,
        max_steps -> BigInt,
        timeout_ms -> BigInt,
        error -> Nullable<Text>,
        summary -> Nullable<Text>,
        source -> Nullable<Text>,
        source_chat_id -> Nullable<BigInt>,
        mode_reasoning -> Nullable<Text>,
        mode_source -> Text,
        messages -> Text,
    }
}

diesel::table! {
    policy (id) {
        id -> Integer,
        capabilities -> Text,
        theme_mode -> Text,
        language -> Text,
        provider_active -> Text,
        provider_model -> Nullable<Text>,
        agent_prompt_append -> Text,
    }
}

diesel::table! {
    schedules (id) {
        id -> Text,
        prompt -> Text,
        capabilities -> Text,
        mode -> Text,
        frequency -> Text,
        cron_expr -> Text,
        enabled -> Bool,
        last_run_at -> Nullable<BigInt>,
        next_run_at -> BigInt,
        created_at -> BigInt,
    }
}

diesel::table! {
    projects (id) {
        id -> Text,
        name -> Text,
        color -> Text,
        created_at -> BigInt,
        task_ids -> Text,
    }
}

diesel::table! {
    channels (id) {
        id -> Text,
        enabled -> Bool,
        status -> Text,
        paired -> Bool,
        pairing_code -> Nullable<Text>,
        error -> Nullable<Text>,
        has_credentials -> Bool,
        meta -> Text,
    }
}

diesel::table! {
    provider_credentials (provider_id) {
        provider_id -> Text,
        api_key -> Text,
        updated_at -> BigInt,
    }
}

diesel::allow_tables_to_appear_in_same_query!(
    tasks,
    policy,
    schedules,
    projects,
    channels,
    provider_credentials,
);
